### ISSUE TITLE
importKey not safe to "derive" key

### ISSUE THREAD

**Issue by fkooman** [filesender/filesender#338]
https://github.com/filesender/filesender/blob/545b987a2a8f1f2bcf6d2dbd0cab2f49ebde3292/www/js/crypter/crypto_app.js#L27

One should use https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/deriveKey in order to derive a key from a password.

Using "raw" is NOT safe with a string, not even with a (base64|acsii85) generated string! You have to make sure (depending on how many bits the encryption algorithm uses) you have the exact number of bits of entropy you feed to "raw".

TL;DR: use `SubtleCrypto.deriveKey()`.

**Comment by fkooman** [comment]
I didn't actually look at the implementation and example there on MDN on what you need to do exactly. The WebCrypto API is horrible as far as I can tell. What I had in mind: https://download.libsodium.org/doc/password_hashing/the_argon2i_function.html (Example 1: key derivation).

Maybe you can find something similar in JS...

**Comment by fkooman** [comment]
> one minor aside is that changing how the password is imported will need some sort of versioning so that files encrypted with the old direct importKey() will still decrypt.

You could keep this information also in the database together with the file information. Or use a special "marker" at the start of the password, e.g. a character that is not used by your password generation function. That is, assuming the decryption function doesn't know if it is able to successfully decrypt. If it does, you can just try decrypting first with the new method, if that fails fall back to old method (for some time).

**Comment by fkooman** [comment]
The only way out seems to be to use libsodium with argon2i(d), aka the `crypto_pwhash` function. The C API: https://download.libsodium.org/doc/password_hashing/the_argon2i_function.html#key-derivation

The JS implementation: https://github.com/jedisct1/libsodium.js

Note: 
> Keep in mind that in order to produce the same key from the same password, the same algorithm, the same salt, and the same values for opslimit and memlimit have to be used. Therefore, these parameters have to be stored for each user.

In the Filesender case for each file, not for each user.

**Comment by fkooman** [comment]
libsodium has the other benefit of having utilities for encoding/decoding of base64/hex etc in constant time. Also random data generator is available.

**Comment by fkooman** [comment]
https://github.com/filesender/filesender/issues/377
